internal/agent/device/console.go

```diff
@@ -0,0 +1,224 @@
+package device
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+
+	grpc_v1 "github.com/flightctl/flightctl/api/grpc/v1"
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/api_server/agentserver"
+	"github.com/flightctl/flightctl/pkg/log"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/metadata"
+)
+
+type ConsoleController struct {
+	grpcClient grpc_v1.RouterServiceClient
+	log        *log.PrefixLogger
+	deviceName string
+
+	active           bool
+	streamClient     grpc_v1.RouterService_StreamClient
+	lastClosedStream string
+}
+
+func NewConsoleController(grpcClient grpc_v1.RouterServiceClient, deviceName string, log *log.PrefixLogger) *ConsoleController {
+	return &ConsoleController{
+		grpcClient: grpcClient,
+		deviceName: deviceName,
+		log:        log,
+	}
+}
+
+func (c *ConsoleController) Sync(ctx context.Context, desired *v1alpha1.RenderedDeviceSpec) error {
+	c.log.Info("Syncing console status")
+	defer c.log.Info("Finished syncing console status")
+
+	// do we have an open console stream, and we are supposed to close it?
+	if desired.Console == nil {
+		c.log.Info("No desired console")
+		if c.active {
+			if c.streamClient != nil {
+				err := c.streamClient.CloseSend()
+				if err != nil {
+					return fmt.Errorf("failed to close console stream: %w", err)
+				}
+			}
+			c.active = false
+		}
+		return nil
+	}
+
+	// do we have a console stream open? we should keep it open
+	if c.active && c.streamClient != nil {
```

Comment:
What if the session id has changed without closing the previous connection? Is such scenario expected to be supported?
May such scenario happen if a new client attempts to create new console connection?

Reply:
Today this will happen:
0) There is an existing console connection (stream A)
1) Client initiates another console connection (stream B)
2) Device gets annotated and gets notified of stream B, drops stream A, connects to B
3) Because stream A is closed, the close is propagated to the client in 0, the client exits
4) Client in 1 was waiting on stream B, gets the console from the connection that happened on 2


So, the first client exits, the new client has a console.


That will drop the device to the new connection. The old connection get's closed client side which is propagated to the client, which closes.

Reply:
What will make (2) to happen if in line 56 if the session id is not checked?